### PR TITLE
Improve handling of custom objects in BP.insert/replace_bot

### PR DESCRIPTION
**Pull request by Bibo-Joshi**
if an object can't be copied, they're probably something special and don't contain a Bot instance. Therefore, those will now just be skipped.

Note: I noticed the latter while testing the changes on a personal project, where I have locks in some objects that have a custom pickle behaviour but not a custom copy behaivour

### CODE REVIEW

File: telegram/ext/basepersistence.py [hunk 1]
```diff
@@ -146,7 +146,13 @@ def replace_bot(cls, obj: object) -> object:
         if isinstance(obj, (list, tuple, set, frozenset)):
             return obj.__class__(cls.replace_bot(item) for item in obj)
 
-        new_obj = copy(obj)
+        try:
+            new_obj = copy(obj)
+        except TypeError as exc:
+            if 'pickle' in str(exc):
+                return obj
```
Comment (tsnoam) [inline telegram/ext/basepersistence.py:153]: add a warning

**Comment by Bibo-Joshi** [comment]
There's another small bug, that should be fixed in this PR: https://t.me/pythontelegrambotgroup/396509

TL;DR, in `insert_bot` replace `if obj == self.REPLACE_BOT` with `if isinstance(obj, str) and obj == self.REPLACED_BOT:`

File: telegram/ext/basepersistence.py [hunk 2]
```diff
@@ -128,12 +128,12 @@ def set_bot(self, bot: Bot) -> None:
         self.bot = bot
 
     @classmethod
-    def replace_bot(cls, obj: object) -> object:
+    def replace_bot(cls, obj: object) -> object:  # pylint: disable=R0911
```
Comment (tsnoam) [inline telegram/ext/basepersistence.py:131]:
```suggestion
    def replace_bot(cls, obj: object) -> object:
```

File: telegram/ext/basepersistence.py [hunk 3]
```diff
@@ -146,7 +146,18 @@ def replace_bot(cls, obj: object) -> object:
         if isinstance(obj, (list, tuple, set, frozenset)):
             return obj.__class__(cls.replace_bot(item) for item in obj)
 
-        new_obj = copy(obj)
+        try:
+            new_obj = copy(obj)
+        except Exception:
```
Comment (tsnoam) [inline telegram/ext/basepersistence.py:151]:
```suggestion
        except Exception:  # pylint: disable=R0911
```

File: telegram/ext/basepersistence.py [hunk 4]
```diff
@@ -146,7 +146,18 @@ def replace_bot(cls, obj: object) -> object:
...
+            warnings.warn(
+                'BasePersistence.replace_bot caught an error while trying to copy an object. '
+                'Objects that can not be copied will be assumed to not contain a telegram.Bot '
+                'instance and will not be handled further. See the docs of '
+                'BasePersistence.replace_bot for more information.',
+                RuntimeWarning,
```
Comment (tsnoam) [inline telegram/ext/basepersistence.py:157]: Try to make this text shorter.
People will get bored and look away with such a long text...

File: telegram/ext/basepersistence.py [hunk 5]
```diff
@@ -183,12 +194,23 @@ def insert_bot(self, obj: object) -> object:  # pylint: disable=R0911
...
-        if obj == self.REPLACED_BOT:
+        if isinstance(obj, str) and obj == self.REPLACED_BOT:
             return self.bot
         if isinstance(obj, (list, tuple, set, frozenset)):
             return obj.__class__(self.insert_bot(item) for item in obj)
 
-        new_obj = copy(obj)
+        try:
+            new_obj = copy(obj)
+        except Exception:
```
Comment (tsnoam) [inline telegram/ext/basepersistence.py:204]:
```suggestion
        except Exception:  # pylint: disable=R0911
```

File: telegram/ext/basepersistence.py [hunk 6]
```diff
@@ -183,12 +194,23 @@ def insert_bot(self, obj: object) -> object:  # pylint: disable=R0911
...
+            warnings.warn(
+                'BasePersistence.insert_bot caught an error while trying to copy an object. '
+                'Objects that can not be copied will be assumed to not contain a telegram.Bot '
+                'instance and will not be handled further. See the docs of '
+                'BasePersistence.insert_bot for more information.',
+                RuntimeWarning,
+            )
```
Comment (tsnoam) [inline telegram/ext/basepersistence.py:211]: Try to make this text shorter.
People will get bored and look away with such a long text...

**Review by tsnoam — CHANGES_REQUESTED** [review]

File: telegram/ext/basepersistence.py (hunk 2 above)
Comment (Bibo-Joshi) [reply telegram/ext/basepersistence.py:131]: pylint complains about the methods having too many return statements, so moving that line to the `except Exception` will make the test fail …

**Review by Poolitzer — APPROVED** [review]
Looks like what we talked about.
